libutils/threaded_deque.c

```diff
@@ -0,0 +1,641 @@
+            GetErrorStrFromCode(ret));
+        pthread_mutexattr_destroy(&attr);
+        free(new_deque->lock);
+        free(new_deque);
+        return NULL;
+    }
+
+    new_deque->cond_non_empty = xmalloc(sizeof(pthread_cond_t));
+    ret = pthread_cond_init(new_deque->cond_non_empty, NULL);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to initialize thread condition "
+            "(pthread_cond_init: %s)",
+            GetErrorStrFromCode(ret));
+        free(new_deque->lock);
+        free(new_deque->cond_non_empty);
+        free(new_deque);
+        return NULL;
+    }
+
+    new_deque->cond_empty = xmalloc(sizeof(pthread_cond_t));
+    ret = pthread_cond_init(new_deque->cond_empty, NULL);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to initialize thread condition "
+            "(pthread_cond_init: %s)",
+            GetErrorStrFromCode(ret));
+        free(new_deque->lock);
+        free(new_deque->cond_empty);
+        free(new_deque->cond_non_empty);
+        free(new_deque);
+        return NULL;
+    }
+
+    return new_deque;
+}
+
+/**
+  @brief Destroys data in range.
+  @warning Assumes that locks are acquired.
+  @note If start == end, this means that all elements in deque will be
+        destroyed. Since the internal array is circular, it will wrap around
+        when reaching the array bounds.
+  @param [in] deque  Pointer to struct.
+  @param [in] start  Position to start destroying from.
+  @param [in] end    First position to not destroy. Can be same as start to
+                     destroy the entire allocated area.
+  */
+static void DestroyRange(ThreadedDeque *deque, size_t start, size_t end)
+{
+    assert(deque != NULL);
+    if (start > deque->capacity || end > deque->capacity)
+    {
+        Log(LOG_LEVEL_DEBUG,
+            "Failed to destroy ThreadedDeque, index greater than capacity: "
+            "start = %zu, end = %zu, capacity = %zu",
+            start, end, deque->capacity);
+        return;
+    }
+
+    if ((deque->ItemDestroy != NULL) && deque->size > 0)
+    {
+        size_t i = start;
+
+        do
+        {
+            deque->ItemDestroy(deque->data[i++]);
+            i %= deque->capacity;
+        } while (i != end);
+    }
+}
+
+/**
+  @brief Either expands capacity of dedeque, or shifts tail to beginning.
+  @warning Assumes that locks are acquired.
+  @param [in] deque  Pointer to struct.
+  */
+static void ExpandIfNecessary(ThreadedDeque *deque)
+{
+    assert(deque != NULL);
+    assert(deque->size <= deque->capacity);
+
+    if (deque->size == deque->capacity)
+    {
+        if (deque->right <= deque->left)
+        {
+            size_t old_capacity = deque->capacity;
+
+            deque->capacity *= EXPAND_FACTOR;
+            deque->data = xrealloc(deque->data,
+                                   sizeof(void *) * deque->capacity);
+
+            memmove(deque->data + old_capacity, deque->data,
```

Comment:
Aren't the arguments here in a wrong order? Also, is the number of items to be moved really always given by `deque->right`?

Reply:
This case only happens when the right index is lower than the left index, in which case we need to make sure we have a continuous block of memory.

`memmove` takes parameters `dest`, `src`, `size`, in that order.

The first position is the position it will be moved to, which is at the first position of the newly allocated memory, at offset `old_capacity` in `deque->data`.

Reply:
I see. So it's moving the "overflow" items that are at the beginning of the data segment, but are actually at the end (right side) of the queue in the "abstract picture". Could you please add an explanatory comment above this call?